server/src/main/java/org/elasticsearch/index/translog/TranslogWriter.java

```diff
@@ -459,8 +467,20 @@ private synchronized ReleasableBytesReference pollOpsToWrite() {
     private void writeAndReleaseOps(ReleasableBytesReference toWrite) throws IOException {
         try (ReleasableBytesReference toClose = toWrite) {
             assert writeLock.isHeldByCurrentThread();
-            ByteBuffer ioBuffer = DiskIoBufferPool.getIoBuffer();
-
+            final int length = toWrite.length();
+            if (length == 0) {
+                return;
+            }
+            ByteBuffer ioBuffer = DiskIoBufferPool.maybeGetDirectIOBuffer();
+            if (ioBuffer == null) {
+                // not using a direct buffer for writes from the current thread so just write without copying to the io buffer
+                BytesRefIterator iterator = toWrite.iterator();
```

Comment:
Is there a case where we are writing using other than write or flush threads? If so, I wonder if we need to add a test or if this is already covered by existing tests?

Reply:
Jup a bunch of them (probably close to all internal cluster tests), we're e.g. running this on the generic thread during global checkpoint sync or recovery from local.

Reply:
Looks like this comes from https://github.com/elastic/elasticsearch/pull/62513#discussion_r498485409 :grin:

Reply:
Can we somehow make `TranslogTests` randomly do one or the other in `DiskIoBufferPool`? Currently, it looks like all the tests there would only test this case and not the case where we actually have a byte-buffer.

Reply:
@henningandersen sure thing this turned out to be surprisingly easy :) see https://github.com/elastic/elasticsearch/pull/77674/commits/91c9b7c5974fb6059c87270a17f09dd2ae7f9e39

Reply:
It looks like that solution only touches 3 tests. I think there are quite a few more tests that are relevant to ensure we randomly cover both code paths here. For instance `testSyncConcurrently`, `testSyncUpTo` and `testSyncUpToStream` (did not run through all tests though).

Reply:
I updated `testSyncConcurrently` now as well because it was easy. I'm not sure about the others, they would require either change the test thread name dynamically or running them on another thread. Not sure that's worth it when the existing tests + all the integ tests hit this code constantly anyway and should give use lots of randomisation for the write sizes?

Reply:
I think we do not need to update the thread names. I think we can turn `DiskIoBufferPool` into an instance and configure the translog with a specific instance to randomize the behavior (and randomly use the production "singleton" instance)? Or we can let translog have a `Supplier<ByteBuffer>` indirection for this.

Reply:
@henningandersen Finally got around to this, sorry for the delay :) passing a customizable instance of `DiskIoBufferPool` around now that randomizes over thread names to init the instance in both ways randomly.

I think this is good for another look whenever you have a second :)